README
mach - make a chroot
--------------------

0. IMPORTANT
------------
If you're on Fedora Core 2, and want to build for older targets,
MAKE SURE you turn off vdso's by doing

sysctl -w kernel.vdso=0

before using mach.
You can add an entry to /etc/sysctl.conf to make this more permanent.

1. WHAT IS IT ?
---------------
mach allows you to set up clean roots from scratch for any distribution or
distribution variation supported.

This clean build root can be used for several goals :
- making clean packages
- set up chroots for services to run it
- make disk images of clean roots (for example for UML)

mach uses python.
This mach is not to be confused with the microkernel of the same name.
If you can't tell the difference between this mach and that mach from a
cursory look then neither is for you.

Currently, mach works for rpm-based distributions that can work with apt
for rpm - hopefully this will be extended to other types of package manager
as well.

2. WHAT DISTROS ARE SUPPORTED ?
-------------------------------
Included at this moment is the necessary information to set up :

Fedora development
  - with rpm.livna.org

Fedora 21
  - everything
  - updates
  - cloud
  - docker
  - server
  - workstation
  - with FreshRPMS
  - with JPackage
  - with rpmfusion.org

Fedora 8-20
  - base
  - updated
  - with rpm.livna.org
  - with FreshRPMS
  - with JPackage
  - with rpmfusion.org

Fedora 7
  - base
  - updated
  - with rpm.livna.org
  - with FreshRPMS
  - with JPackage

Fedora 4, 5, 6
  - Core
  - updated
  - with rpm.livna.org
  - with FreshRPMS
  - with JPackage

Fedora Development/Rawhide

OpenSuse 10.2, 10.3, 11.0

Fedora 1, 2, 3
  - Core
  - updated
  - with fedora.us stable, testing and unstable/extras
  - with rpm.livna.org stable, testing and unstable
  - with FreshRPMS
  - with JPackage

Red Hat 8.0 and 9
  - standard
  - updated
  - with fedora.us stable, testing and unstable
  - with rpm.livna.org stable, testing and unstable
  - with FreshRPMS
  - with JPackage

Red Hat 7.2 and 7.3
  - standard
  - updated
  - with FreshRPMS
  - with JPackage

Red Hat 7.0 and 7.1
  - standard
  - updated
  - with FreshRPMS

CentOS 3, 4, 5, 6, 7
  - standard
  - updated
  - with addons
  - with contrib (3, 4 only)
  - with csgfs (3, 4 only)
  - with extras
  - with centosplus
  - with Fedora Extras rebuilds (4 and higher only)
  - with EPEL (4 and higher only)
  - with JPackage (4 and higher only)

Dave/Dina
  - fridge
  - oven

SuSE 8.1, 8.2 and 9.0
  - base
  - updated

Yellowdog 2.3 and 3.0
  - standard
  - updated
  - with FreshRPMS

Some handy features of mach include
  - "caching" of downloaded packages using the build hosts's apt
    the build root
  - ensures clean packages by reverting to the base set of build packages
  - uses apt to resolve dependencies
  - parsing of BuildRequires to install necessary packages for building
  - build ordering when doing multiple builds
  - support for flavours of distribution
  - multiple build roots
  - locking of buildroot to avoid concurrent builds
  - optional signing of built packages

3. REQUIREMENTS
---------------
- A Linux kernel that can do bind mounts (>= 2.2.0 if I'm correct)
- rpm (>= 4.1 + gpg and cpio if you want GPG)
- python
- apt >= 0.5.5cnc2 or yum
- sudo
- lots of disk space
- good faith

4. GETTING STARTED - RPM INSTALL
--------------------------------
- The easiest way to install mach is to just install the RPM.
- After installing, you need to add all users that are going to build through
  mach to the "mach" group in /etc/group, and (possibly) log out and log back
  in so that this change gets registered.  You can check if your user
  is in this group by trying "groups (user)"

5. GETTING STARTED - SOURCE INSTALL
-----------------------------------
- If you insist on running from source, do
  ./configure
  make
  make install

  as root.

- you have to setuid the mach-helper binary.
  chmod u+s $(prefix)/sbin/mach-helper

- do not start to tweak stuff all in the source code just yet; mach is picky
  about its setup if you start to hack it.

6. SETTING IT UP
----------------
- configuration files are stored in $(sysconfdir)/mach
  dist contains the list of packages for each distro target, as well
  as download locations for various packages.
  conf contains configuration data, like hosts and resolv.conf files

- It is advised you do not hand-edit these files.  Instead, you can
  override any configuration variable by putting it in
  ~/.machrc

- you can choose between using apt and yum by setting
  config['installer'] = 'yum'
  or
  config['installer'] = 'apt-get'
  in .machrc.

  Each time you change this, you will have to clean up your roots and
  repopulate them.

- you can add dist-specific configurations by creating a dict in ~/.machrc
  It is advised that you create at least a files dict containing
  /etc/resolv.conf and /etc/hosts.

  For example, I have in my .machrc:

config['files'] = {
  '/etc/hosts': '''
127.0.0.1 mach.onshuis onzenbak
''',
  '/etc/resolv.conf': '''
nameserver 127.0.0.1
nameserver 157.193.40.42
search onshuis
'''
}

config['script-success'] = "mach-collect"

This ensures that all packages appear to be built on the host mach.onshuis,
and ensures that it can look up hosts through localhost DNS and one outside
DNS.
Also, after a successful build it will run the mach-collect script with
the built directory as a parameter.

Here is an example of a mach-collect script:

#!/bin/bash
pwd
ls $1
mkdir -p RPMS
cp $1/*.rpm RPMS

This script creates an RPMS dir in the current directory, and copies all .rpm
files there.

7. USING IT
-----------

- If you or your users want to have a private own personal root, then insert
  these lines into the .machrc file:

    config['dirs']['states'] += "/%s" % os.environ.get("USER")
    config['dirs']['roots'] += "/%s" % os.environ.get("USER")
    config['dirs']['results'] += "/%s" % os.environ.get("USER")
    config['dirs']['tmps'] += "/%s" % os.environ.get("USER")

- run mach --help for a list of commands you can use.
- Now for the actual work.  If everything is configured properly, you can now
  do
    mach setup base
  and a base root for the default distribution should be created for you.
  run
    mach chroot
  and poke around a little.

- To build a package based on a .spec file, run
  mach build (spec file name)

  If the Source: and Patch: lines contain a proper URL, mach will automatically
  download them and build the package.

- To build a package from a .src.rpm file, run
  mach rebuild (src.rpm)
  You can specify an http:// or ftp:// file here, and it will be downloaded
  and built automatically.

- If the build fails because it needs other packages to build and they weren't
  mentioned in the spec file, you can either:
  - change the spec file and add the BuildRequires:
  - install the packages through mach
       mach apt-get -y install (requires)
    and run the build again, with the -k option (for "keep packages")

- You can build multiple spec files or src.rpms, just try
       mach build *.spec
  or
       mach build *.src.rpm

- To clean the buildroot, run
    mach clean

8. SECURITY
-----------
- By nature mach needs root privileges for some of its operations.
  All operations that it needs this for are done through the mach-helper
  binary, which is suid.  The source code for this binary is deliberately
  kept both simple and very minimal.  The goal here is to make it absolutely
  clear what mach can and cannot do as root.
- If you find any security problems with this, PLEASE let me know.  I don't
  have a lot of experience with secure programming and can use help there.

9. TIPS & TRICKS
----------------
- mach tries to adhere to the FHS.  It uses a lot of disk space by nature.
  You might not have as much space as you would like.  Since security needs
  to be tight, mach-helper is very strict about this location.  A good
  solution for this is to use bind mounts.  For example, I have in my
  /etc/fstab file:
  /home/mach/roots /var/lib/mach/roots auto bind,defaults 0 0
  which causes /home/mach/roots on my big home disk to be mounted again
  on /var/lib/mach/roots.

- mach uses the host's rpm binary to manipulate the database inside the root.
  A lot of fiddling went into this; I could find of no other way to guarantee
  it works reliably for all target platforms.
  As a side effect, you will have to manipulate the target's rpm database
  using rpm and apt from outside the host.
  The easiest way to do this is to use mach to operate on it; for example,
  mach rpm -Uhv (rpm) will install the given rpm in the target's system.

- to build kernel modules or other packages that BuildRequire the kernel
  or kernel-source rpms, you need to manually install the kernel and
  kernel-source packages with the correct version.
  As an example:
  mach -r f1fs apt-get -y install kernel-source#2.4.22-1.2115.nptl
  to install release 1.2115.ntpl of version 2.4.22
  It is best to first try building your .spec file so that the buildroot
  gets restored.  After that, install the kernel packages.  Then, rerun
  the build with -k (to keep packages in the buildroot) and the build should
  succeed.

- SELINUX:
  When using an SELINUX-enabled host, a whole range of things can fail.
  For example, creating an FC2 target on an FC3 host fails when installing
  packages.
  I've added a library that gets preloaded to override selinux, which allows
  most packages to be installed.

10. BUGS
--------
- rpm 4.1 included with Red hat 8.0 has a few race condition bugs which
  can cause rpm to hang.  If this happens, you should kill -9 the stalling
  rpm process, chroot into the root, run "rm -rf /var/lib/rpm/__*" and retry.
  had given.
  I'm interested on feedback running mach on RH8.0, I have no such system
  around anymore.
- for now, report them to me personally at thomas (at) apestaart (dot) org
- mounting /proc into the root system can cause problems.  For example,
  removing openssh-server from the root will cause it to kill your build host's
  ssh server, since it gets the pid from proc.  This isn't what you want to
  have happen if the server is not somewhere in your physical neighbourhood.
  This is the reason why openssh-server is listed in the base install.

11. MAILING LIST
----------------
A mailing list has been set up for discussion of mach use and development.
Check http://lists.sourceforge.net/lists/listinfo/mach-devel for information.
The list is low-volume.

12. ADDING A DISTRIBUTION
-------------------------
- Check the distribution files in $sysconfdir/dist.d and copy one.
- Make sure the download locations you reference are specified in location.

- Base packages: list them in the packages[dist] dict under 'base'.
  You can then test by trying to set up this root,
  See if the installation of base packages works well by doing
  mach -r (dist) setup base

13. DISCLAIMER
--------------
mach is offered under the GPL without any warranty whatsoever.
Let me know where the documentation is lacking, so I can improve on that.


NEWS
This is mach, make a chroot, 1.0.4.

Changes since 1.0.3:
	- Added Fedora 21 and CentOS 7.

Changes since 1.0.2:
	- Added Fedora 19 and 20.

Changes since 1.0.1:
	- Added Fedora 18.
	- Fixed bug for build.

Changes since 1.0.0:
	- Added Fedora 17.
        - Fixed Fedora development.
	- Fixed bug for build.

Changes since 1.0.0:
	- Added Fedora 15, 16.
	- Added CentOS 6.

Changes since 0.9.5:
	- Added Fedora 12, 13, 14

Changes since 0.9.4:
	- Added Fedora 10 and 11
	- Change to use archived Fedora versions.
	- Fix for more users in mach group.
        - Fixes for python 2.6

Changes since 0.9.3:
	- Add fix for yum's new --version output breaking mach (Thomas)
	- Fix permission problems for additional users (Julien)
	- Add OpenSuse 10.2, 10.3, 11.0 build files (Julien)

Changes since 0.9.2:
        - Add F8 and F9 repositories and reorganize repositories (Thomas, Ville)
        - Default to CentOS configs when built on RHEL (Ville)
	- Remove Connectiva 9 (Ville)

Changes since 0.9.1:
        - Add FC7, CentOS EPEL, CentOS 3 configs (Ville)
	- Allow overriding the pkg dir under /usr/src (Tobias)
	- Clarify GPLv2+ license (Ville)
	- Run rpmlint on all result dirs (Ville)
	- Actually include ppc files (Thomas)

Changes since 0.9.0:
        - Add FC6, CentOS 4 and JPackage 1.7 configs (Ville)
        - Run rpmlint by default (Thomas)
        - Cleanup of dist files (Ville)
        - Enable buildroot and rpath checks in FE roots (Ville)
	- Add /dev/(u)random to the root (Thomas)
	- Sync FC build package sets with Fedora guidelines (Ville)
        - Create RPM transaction lock directories during prep (Nigel)
	- Add Fedora PPC configs (Matthias)
        - Add debian build files (Jan)

Changes since 0.4.9:
        - Fix livna.org and rawhide URL's (Ville)
        - Fix "revert to build list" such that it can deal with
          upgraded versions of the build packages, as well as
          deal with packages with broken post/postun scripts (Thomas)
        - make sure we refresh local repository cache with
          yum >= 2.6.0 (Nigel)
        - always recreate repository config files, so that changes
          to "location" are effective immediately (Thomas)
        - Add JPackage URL's for FC3, 4, 5 (Ville)
        - mach clean also cleans out the local repo (Thomas)

Changes since 0.4.8:
        - added DOAP file (Thomas)
        - added Fedora Core 5 (Thomas)
        - fix build ordering when building multiple packages (Thomas)
        - always evaluate spec file inside the root (Thomas)
	- add an option for md5summing of results (Matthias)
        - pass 'buildopts' from a root config to rpmbuild (Thomas)
        - add 'excludearch' property to root config,
          fixes yum's semi-random behaviour on x86_64 (Thomas)
        - store yum cache data per-dist, shared for flavors (Thomas)
        - clean up yum metadata for local repositories (Thomas)

Changes since 0.4.7:
        - multi-user/group-related fixes
        - fixes for FC3 running on FC4 host

Changes since 0.4.6:
        - add yum support (various people)
        - Work with SELinux enabled (Thomas)
        - Remove all matches when removing packages (Rudi)
        - Better config error handling (Ville)
        - Fix filename munging on collect (Jeff)
        - add support for FC3 and FC4 (Thomas)
        - Create /dev/null properly (Rudi)
        - Add runuser support (Rudi)
        - support fedora extra's "dist" var (Thomas)
        - Fix urlgrab exceptions (Ville)
        - upgrade packages before snapshotting build list (Thomas)

Changes since 0.4.4:
	- pre-interpreting checks fixed (Thomas)
	- use promoteepoch when supported; fixes building with newer apt
          for older platforms (Thomas)
        - proc mount/unmount fixes (Erik, Enrico)
	- distro configuration split up in multiple distro files (Thomas)
	- rewrite macros on each build (Thomas)
	- pass through a specific set of env vars (Enrico)
	- create /boot/kernel.h on root setup (Enrico)
        - fix handling of rpmmacros when checking spec files (Thomas)
	- always use build args to parse spec flies (Thomas)
        - added -v option (Thomas)

Changes since 0.4.3:
	- added distributions:
		- JPackage for FC1 (Ville)
		- Conectiva Linux 9 (Ville)
		- RH80/RH9/FC1 with rpm.livna.org (Thomas)
	- fixed distributions:
		- Red Hat 7.3 (Thomas)
	- make rpmbuild command configurable (Ville)
	- use config['macros'] to allow additional macros (Enrico)
	- fix permissions on mach directories for multiple users (Thomas)
	- add "script-success" to run a script on succesful build (Thomas)
	- fix quoting problems so --target and --define work (Thomas)
	- warn on kernel-related BuildRequires (Thomas)
	- verify parsibility of spec file (Thomas)
	- recreate sources.list on each update (Thomas)
	- autodetect target distro on configure (Thomas)
	- revert to package snapshot list between builds (Thomas)
	- add pre-build in-root buildrequires verification (Thomas)

Changes since 0.4.2:
	- added support for Fedora 1
	- moved config files per distro initialization to dist file

Changes since 0.4.1:
	- added support for Fedora Core 0.95 (Severn beta 3)
	- moved config files per distro initialization to dist file

Changes since 0.4.0:
	- added Red Hat 7.0, 7.1
	- added Yellowdog 2.3, 3.0
	- added Fedora Core 0.94
	- added FreshRPMS targets
	- make "mach apt-get" run interactive
	- implemented "mach status"
	- fix permissions on copied sources
	- unlock root on successful build
	- allow "." in mach-helper


src/selinux-mach.c
/*
 * Copyright (C) 2003-2007 Thomas Vander Stichele
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include "config.h"
#include <selinux/selinux.h>

extern int is_selinux_enabled(void)
{
  /* always return 0; this way we don't trigger any SELINUX calls */
  return 0;
}

/* this function gives failures when installing basic rpms in the root;
 * so we fake it out as well */
#if HAVE_LSETFILECON_TYPED
extern int lsetfilecon(const char *path, security_context_t con)
#else
extern int lsetfilecon(const char *path, const char *con)
#endif
{
  return 0;
}

int rpm_execcon(unsigned  int  verified,  const  char  *filename,  char
       *const argv[] , char *const envp[]);



src/mach-helper.c
/*
 * mach-helper.c: help mach perform tasks needing root privileges
 *
 * Copyright (C) 2003-2007 Thomas Vander Stichele
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#define _GNU_SOURCE

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <errno.h>
#include <stdarg.h>
#include <stdlib.h>
#include <string.h>

/* pull in configure'd defines */
char *rootsdir = ROOTSDIR;
char *statesdir = STATESDIR;
char *archivesdir = ARCHIVESDIR;
char *cachedir = LOCALSTATEDIR "/cache";

static char const * const ALLOWED_ENV[] =
{
  "APT_CONFIG",
  "dist",
  "ftp_proxy", "http_proxy", "https_proxy", "no_proxy"
};

#define ALLOWED_ENV_SIZE (sizeof (ALLOWED_ENV) / sizeof (ALLOWED_ENV[0]))

/*
 * helper functions
 */

void
usage ()
{
  printf ("Usage: mach-helper [command]\n");
  exit (1);
}

/* print formatted string to stderr, print newline and terminate */
void
error (const char *format, ...)
{
  va_list ap;

  va_start (ap, format);
  fprintf (stderr, "mach-helper: error: ");
  vfprintf (stderr, format, ap);
  va_end (ap);
  fprintf (stderr, "\n");
  exit (1);
}

/*
 * perform checks on the given dir
 * - is the given dir under the allowed hierarchy ?
 * - is it an actual dir ?
 * - are we not being tricked by using . or .. ?
 */
void
check_dir_allowed (const char *allowed, const char *given)
{
  struct stat buf;
  char last;
  int retval;

  /* does given start with allowed ? */
  if (strncmp (given, allowed, strlen (allowed)) != 0)
    error ("%s: not under allowed directory (%s)", given, allowed);

  /* does it try to fool us by using .. ? */
  if (strstr (given, "..") != 0)
    error ("%s: contains '..'", given);

  /* does it try to fool us into following symlinks by having a trailing / ? */
  last = given[strlen (given) - 1];
  if (last == '/')
    error ("%s: ends with '/'", given);

  /* are we chrooting to an actual directory (not a symlink or anything) ? */
  retval = lstat (given, &buf);
  if (retval != 0)
    error ("%s: %s", given, strerror (errno));

  //printf ("DEBUG: mode: %o\n", buf.st_mode);
  if (S_ISLNK (buf.st_mode))
    error ("%s: symbolic link", given);
  if (!(S_ISDIR (buf.st_mode)))
    error ("%s: not a directory", given);
}

/*
 * perform checks on the given file
 * - is the given file under the allowed hierarchy ?
 * - is it an actual file ?
 * - are we not being tricked by using .. ?
 */
void
check_file_allowed (const char *allowed, const char *given)
{
  struct stat buf;
  char last;
  int retval;

  /* does given start with allowed ? */
  if (strncmp (given, allowed, strlen (allowed)) != 0)
    error ("%s: not under allowed directory", given);

  /* does it try to fool us by using .. ? */
  if (strstr (given, "..") != 0)
    error ("%s: contains '..'", given);

  /* does it have a trailing / ? */
  last = given[strlen (given) - 1];
  if (last == '/')
    error ("%s: ends with '/'", given);

  /* are we working with an actual file ? */
  retval = lstat (given, &buf);
  if (retval != 0)
    error ("%s: %s", given, strerror (errno));

  //printf ("DEBUG: mode: %o\n", buf.st_mode);
  if (S_ISLNK (buf.st_mode))
    error ("%s: symbolic link", given);
  if (!(S_ISREG (buf.st_mode)))
    error ("%s: not a regular file", given);
}

/* argv[0] should by convention be the binary name to be executed */
void
do_command (const char *filename, char *const argv[])
{
  /* do not trust user environment;
   * copy over allowed env vars, after setting PATH and HOME ourselves
   */
  char *env[3 + ALLOWED_ENV_SIZE] = {
    [0] = "PATH=/bin:/usr/bin:/usr/sbin",
    [1] = "HOME=/root"
  };
  size_t idx=2;
  size_t i;
  char *envvar;
  char *ld_preload;
  struct stat buf;

  /* elevate privileges */
  if (setreuid (geteuid (), geteuid ()) != 0)
    error ("setreuid: %s", strerror (errno));
  //printf ("DEBUG: First argument: %s\n", *argv);
  //printf ("DEBUG: Executing %s\n", filename);
  /* FIXME: for a debug option */
  /*
  printf ("Executing %s ", filename);
  char **arg;
  for (arg = (char **) &(argv[1]); *arg; ++arg)
    printf ("%s ", *arg);
  printf ("\n");
  */

  /* add LD_PRELOAD for our selinux lib if MACH_LD_PRELOAD is set, or if
   * /selinux exists */
  envvar = getenv ("MACH_LD_PRELOAD");
  if (envvar != 0 || stat ("/selinux", &buf) == 0)
  {
    ld_preload = strdup("LD_PRELOAD=" LIBDIR "/libselinux-mach.so");
    env[idx++] = ld_preload;
  }

  for (i = 0; i < ALLOWED_ENV_SIZE; ++i)
  {
    char *ptr = getenv (ALLOWED_ENV[i]);
    if (ptr==0) continue;
    ptr -= strlen (ALLOWED_ENV[i]) + 1;
    env[idx++] = ptr;
  }

  execve (filename, argv, env);
  error ("executing %s: %s", filename, strerror (errno));
}

/*
 * actual command implementations
 */

/* allow apt-get -c (statesdir/.../apt.conf) ... */
void
do_apt_get (int argc, char *argv[])
{
  /* enough arguments ? mach-helper apt-get -c (statesdir/.../apt.conf) .., 5 */
  if (argc < 5)
    error ("not enough arguments");

  /* -c */
  if (strncmp ("-c", argv[2], 2) != 0)
    error ("%s: options not allowed", argv[2]);

  /* check given file */
  check_file_allowed (statesdir, argv[3]);

  do_command ("/usr/bin/apt-get", &(argv[1]));
}

/* allow yum -c (statesdir/.../yum.conf) ... */
void
do_yum (int argc, char *argv[])
{
  char *installroot = NULL;

  /* enough arguments ?
   * mach-helper yum -c (statesdir/.../yum.conf) --installroot=..., 6 */
  if (argc < 6)
    error ("not enough arguments");

  /* -c */
  if (strncmp ("-c", argv[2], 2) != 0)
    error ("%s: options not allowed", argv[2]);

  /* check given file */
  check_file_allowed (statesdir, argv[3]);

  /* installroot */
  if (strncmp ("--installroot=", argv[4], 14) != 0)
    error ("%s: option not allowed", argv[4]);

  installroot = argv[4] + 14;
  check_dir_allowed (rootsdir, installroot);

  do_command ("/usr/bin/yum", &(argv[1]));
}

void
do_chroot (int argc, char *argv[])
{
  if (argc < 3)
    error ("No directory given for chroot !");
  //printf ("DEBUG: rootsdir: %s\n", rootsdir);

  /* do we allow this dir ? */
  check_dir_allowed (rootsdir, argv[2]);

  do_command ("/usr/sbin/chroot", &(argv[1]));
}

/*
 * allow bind mounts of the archives or selinux dir:
 * mount -o bind (archivesdir) (root)/var/cache/apt/archives
 * mount -o bind /selinux (root)/selinux
 * allow proc mounts:
 * mount -t proc proc (root)/proc
 */
void
do_mount (int argc, char *argv[])
{
  /* see if we have enough arguments for it to be what we want, ie. 5 */
  if (argc < 5)
    error ("not enough arguments");

  /* see if it's -o bind or -t proc */
  if ((strncmp ("-o", argv[2], 2) == 0) && (strncmp ("bind", argv[3], 4) == 0))
  {
    /* see if it's the archivesdir or selinux */
    if ((strncmp (archivesdir, argv[4], strlen (archivesdir)) != 0) &&
        (strncmp ("/selinux", argv[4], strlen ("/selinux")) != 0))
    {
      error ("%s: mount not allowed", argv[4]);
    }
  }
  else if ((strncmp ("-t", argv[2], 2) == 0) &&
           (strncmp ("proc", argv[3], 4) == 0))
  {
    /* see if we're mounting proc to somewhere in rootsdir */
    if (strncmp (rootsdir, argv[5], strlen (rootsdir)) != 0)
      error ("proc: mount not allowed on %s", argv[5]);
  }
  else
    error ("unallowed mount type");

  /* all checks passed, execute */
  do_command ("/bin/mount", &(argv[1]));
}

void
do_rm_cache_file (int argc, char *argv[])
{
  /* see if we're doing rm on a file under cachedir */
  check_file_allowed (cachedir, argv[2]);

  /* all checks passed, execute */
  do_command ("/bin/rm", &(argv[1]));
}

void
do_rm_cache_dir (int argc, char *argv[])
{
  /* see if we're doing rm -rfv */
  if (strncmp ("-rfv", argv[2], 4) != 0)
    error ("%s: options not allowed", argv[2]);

  /* see if we're doing -rfv on a dir under rootsdir */
  check_dir_allowed (cachedir, argv[3]);

  /* all checks passed, execute */
  do_command ("/bin/rm", &(argv[1]));
}

void
do_rm_root_dir (int argc, char *argv[])
{
  /* see if we're doing rm -rfv */
  if (strncmp ("-rfv", argv[2], 4) != 0)
    error ("%s: options not allowed", argv[2]);

  /* see if we're doing -rfv on a dir under rootsdir */
  check_dir_allowed (rootsdir, argv[3]);

  /* all checks passed, execute */
  do_command ("/bin/rm", &(argv[1]));
}

/* clean out a chroot dir, or remove a mach cache file
 * rm cachedir/mach/somefile
 * rm -rfv (cachedir)
 * rm -rfv (rootdir)
 */


void
do_rm (int argc, char *argv[])
{
  if (argc == 3) {
    do_rm_cache_file (argc, argv);
    return;
  }

  if (argc != 4)
      error ("wrong number of parameters");

  if (strncmp (argv[3], rootsdir, strlen (rootsdir)) == 0)
    do_rm_root_dir (argc, argv);
  else if (strncmp (argv[3], cachedir, strlen (cachedir)) == 0)
    do_rm_cache_dir (argc, argv);
  else
    error ("%s: not under allowed directories", argv[3]);
}

/* perform rpm commands on root */
void
do_rpm (int argc, char *argv[])
{
  const char *rpm_bin = NULL;
  struct stat buf;

  /* enough arguments ? mach-helper rpm --root (rootdir) ... , 4 */
  if (argc < 4)
    error ("not enough arguments");

  /* --root */
  if (strncmp ("--root", argv[2], 6) != 0)
    error ("%s: options not allowed", argv[2]);

  /* Find our rpm binary */
  if (stat ("/bin/rpm", &buf) == 0)
    rpm_bin = "/bin/rpm";
  else if (stat ("/usr/bin/rpm", &buf) == 0)
    rpm_bin = "/usr/bin/rpm";
  else
    error ("Could not locate rpm binary in /bin or /usr/bin");

  /* check given dir */
  check_dir_allowed (rootsdir, argv[3]);

  /* all checks passed, execute */
  do_command (rpm_bin, &(argv[1]));
}


/* unmount archivesdir and proc */
void
do_umount (int argc, char *argv[])
{
  /* enough arguments ? mach-helper umount (dir), 3 */
  if (argc < 3)
    error ("not enough arguments");

  /* see if we're unmounting from somewhere in rootsdir */
  check_dir_allowed (rootsdir, argv[2]);

  /* all checks passed, execute */
  do_command ("/bin/umount", &(argv[1]));
}

/* make /dev/null or /dev/zero device node */
void
do_mknod (int argc, char *argv[])
{
  /* enough arguments ? mach-helper mknod (name) -m (mode) (type) (major) (minor), 8 */
  if (argc < 8)
    error ("not enough arguments");

  /* check given file */
  if (strncmp (argv[2], rootsdir, strlen (rootsdir)) != 0)
    error ("%s: not under allowed directory (%s)", argv[2], rootsdir);

  /* does it try to fool us by using .. ? */
  if (strstr (argv[2], "..") != 0)
    error ("%s: contains '..'", argv[2]);

  /* does it have a trailing / ? */
  int last = argv[2][strlen (argv[2]) - 1];
  if (last == '/')
    error ("%s: ends with '/'", argv[2]);

  /* -m */
  if (strncmp ("-m", argv[3], 2) != 0)
    error ("%s: options not allowed", argv[3]);

  /* type */
  if (strncmp ("666", argv[4], 3) != 0)
    error ("%s: options not allowed", argv[4]);

  /* type */
  if (strncmp ("c", argv[5], 1) != 0)
    error ("%s: options not allowed", argv[5]);

  /* major */
  if (strncmp ("1", argv[6], 1) != 0)
    error ("%s: options not allowed", argv[6]);

  /* minor */
  if (strlen(argv[7]) != 1)
    error ("%s: options not allowed", argv[7]);

  if (*argv[7] != '3' && *argv[7] != '5' && *argv[7] != '8' && *argv[7] != '9')
    error ("%s: options not allowed", argv[7]);

  /* all checks passed, execute */
  do_command ("/bin/mknod", &(argv[1]));
}

/* allow showing the environment, but nothing else */
void
do_env (int argc, char *argv[])
{
  /* enough arguments ? mach-helper mknod (name) -m (mode) (type) (major) (minor), 8 */
  if (argc > 2)
    error ("too many arguments");

  do_command ("/bin/env", &(argv[1]));
}

int
main (int argc, char *argv[])
{

  /* verify input */
  if (argc < 2) usage ();

  /* see which command we are trying to run */
  if (strncmp ("chroot", argv[1], 6) == 0)
    do_chroot (argc, argv);
  else if (strncmp ("mount", argv[1], 5) == 0)
    do_mount (argc, argv);
  else if (strncmp ("rm", argv[1], 2) == 0)
    do_rm (argc, argv);
  else if (strncmp ("umount", argv[1], 6) == 0)
    do_umount (argc, argv);
  else if (strncmp ("rpm", argv[1], 3) == 0)
    do_rpm (argc, argv);
  else if (strncmp ("apt-get", argv[1], 7) == 0)
    do_apt_get (argc, argv);
  else if (strncmp ("yum", argv[1], 3) == 0)
    do_yum (argc, argv);
  else if (strncmp ("mknod", argv[1], 5) == 0)
    do_mknod (argc, argv);
  else if (strncmp ("env", argv[1], 3) == 0)
    do_env (argc, argv);
  else
  {
    error ("Command %s not recognized !\n", argv[1]);
    exit (1);
  }
  exit (0);
}
